### PR TITLE
EI-279 Refresh VRF case status for legal entities - timer triggered function

### PR DESCRIPTION
**Pull request by ruslanss**
…unction

### CODE REVIEW

File: src/SFA.DAS.EmployerIncentives.Functions.AcceptanceTests/Steps/VendorRegistrationFormSteps.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        {
+            var lastRunDate = await _configuration.GetLastRunDateTime();
+
+            _testContext.EmployerIncentivesApi.MockServer
+                 .Given(
+                     Request
+                         .Create()
+                         .WithPath("/api/legalentities/vendorregistrationform/status")
+                         .WithParam("from", new ExactMatcher($"{lastRunDate:yyyyMMddHHmmss}"))
+                         .WithParam("to", new ExactMatcher("20200901020304"))
```
Comment (ruslanss) [inline src/SFA.DAS.EmployerIncentives.Functions.AcceptanceTests/Steps/VendorRegistrationFormSteps.cs:39]: i thought this is well worth asserting on?

**Review by David-B-Read — APPROVED** [review]

File: src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+            IDateTimeProvider dateTimeProvider)
+        {
+            _vendorRegistrationFormService = vendorRegistrationFormService;
+            _configuration = configuration;
+            _dateTimeProvider = dateTimeProvider;
+        }
+
+        // Every hour: "* 0 * * * *"
+        [FunctionName("RefreshVendorRegistrationCaseStatus")]
+        public async Task Run([TimerTrigger("%RefreshVendorRegistrationCaseStatusTriggerTime%")] TimerInfo myTimer, ILogger log)
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs:25]: Inject the logger in the constructor.

File: src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            var from = await _configuration.GetLastRunDateTime();
+            var to = await _dateTimeProvider.GetCurrentDateTime();
+
+            log.Log(LogLevel.Information, $"Calling RefreshVendorRegistrationFormStatuses with parameters: [dateTimeFrom={from}] & [dateTimeTo={to}]");
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs:30]: Do we need this?

File: src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+            await _vendorRegistrationFormService.RefreshVendorRegistrationFormStatuses(from, to);
+
+            await _configuration.UpdateLastRunDateTime(to);
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs:34]: Shouldn't this be stored in the domain and we just call refresh?  Then the API would know what needs to be refreshed and the date check is in the employer incentives domain?

**Review by Stephek — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/VrfCaseRefreshConfiguration.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.EmployerIncentives.Functions.LegalEntities
+{
+    public class VrfCaseRefreshConfiguration : IVrfCaseRefreshConfiguration
+    {
+        private static readonly DateTime DefaultLastRunDateTime = new DateTime(2020, 8, 1);
+        private readonly CloudTable _table;
+        private const string PartitionKey = "LOCAL";
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/VrfCaseRefreshConfiguration.cs:14]: This is environment specific

**Review by Stephek — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+            _vendorRegistrationFormService = vendorRegistrationFormService;
+            _configuration = configuration;
+            _dateTimeProvider = dateTimeProvider;
+        }
+
+        // Every hour: "* 0 * * * *"
+        [FunctionName("RefreshVendorRegistrationCaseStatus")]
+        public async Task Run([TimerTrigger("%RefreshVendorRegistrationCaseStatusTriggerTime%")] TimerInfo myTimer, ILogger log)
+        {
+            var from = await _configuration.GetLastRunDateTime();
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs:27]: Would it be better to put GetLastRunDateTime, GetCurrentDateTime and UpdateLastRunDateTimein the service?  Then at least the function can be dumb as it should be and the logic is contained in the service?

Then you can decorate the service to add the logging and keep the concerns separate.

File: src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs (hunk 4 above)
Comment (Stephek) [reply src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs:34]: Spoke to Mike and he said this was the agreed approach.

**Review by Stephek — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs [hunk 7]
```diff
@@ -0,0 +1,37 @@
+﻿using Microsoft.Azure.WebJobs;
+using Microsoft.Extensions.Logging;
+using SFA.DAS.EmployerIncentives.Functions.LegalEntities.Services.LegalEntities;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.EmployerIncentives.Functions.LegalEntities
+{
+    public class RefreshVendorRegistrationCaseStatus
+    {
+        private readonly IVendorRegistrationFormService _vendorRegistrationFormService;
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs:10]: "Form" service?  Isn't it just the VendorRegistrationService  ?

**Review by Stephek — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/Services/LegalEntities/IVendorRegistrationFormService.cs [hunk 8]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.EmployerIncentives.Functions.LegalEntities.Services.LegalEntities
+{
+    public interface IVendorRegistrationFormService
+    {
+        Task RefreshVendorRegistrationFormStatuses(DateTime fromDateTime, DateTime toDateTime);
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/Services/LegalEntities/IVendorRegistrationFormService.cs:8]: Can this be re-named to RefreshStatuses ?  It's on the IVendorRegistrationFormService so we know it applies to that.

**Review by Stephek — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs (hunk 2 above)
Comment (ruslanss) [reply src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs:25]: this is a standard practice as per microsoft's documentation. logger here gets injected automagically and it's consistent with how this function was done before  :)

File: src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs (hunk 3 above)
Comment (ruslanss) [reply src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs:30]: yes

File: src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/VrfCaseRefreshConfiguration.cs (hunk 5 above)
Comment (ruslanss) [reply src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/VrfCaseRefreshConfiguration.cs:14]: checking with devops

File: src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs (hunk 6 above)
Comment (ruslanss) [reply src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs:27]: not really

File: src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs (hunk 7 above)
Comment (ruslanss) [reply src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs:10]: VRF

Comment (ruslanss) [reply src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs:10]: VRF Service

File: src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs (hunk 2 above)
Comment (Stephek) [reply src/SFA.DAS.EmployerIncentives.Functions.LegalEntities/RefreshVendorRegistrationCaseStatus.cs]: That's before .net core supported DI.

**Review by Stephek — APPROVED** [review]

**Pull request merged**